server/vulnerabilities/cpe.go

```diff
@@ -140,67 +115,391 @@ func cleanAppName(appName string) string {
+		release, err := GetLatestNVDRelease(client)
+		if err != nil {
+			return err
+		}
+		stat, err := os.Stat(path)
+		switch {
+		case errors.Is(err, os.ErrNotExist):
+			// okay
+		case err != nil:
+			return err
+		default:
+			if stat.ModTime().After(release.CreatedAt.Time) {
+				// file is newer than release, do nothing
+				return nil
+			}
+		}
+
+		for _, asset := range release.Assets {
+			if cpeTranslationsFilename == asset.GetName() {
+				cpeTranslationsURL = asset.GetBrowserDownloadURL()
+				break
+			}
+		}
+		if cpeTranslationsURL == "" {
+			return errors.New("failed to find cpe translations in nvd release")
+
+		}
+	}
+
+	u, err := url.Parse(cpeTranslationsURL)
+	if err != nil {
+		return err
+	}
+	if err := download.Download(client, u, path); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// regexpCache caches compiled regular expressions. Not safe for concurrent use.
+type regexpCache struct {
+	re map[string]*regexp.Regexp
+}
+
+func newRegexpCache() *regexpCache {
+	return &regexpCache{re: make(map[string]*regexp.Regexp)}
+}
+
+func (r *regexpCache) Get(pattern string) (*regexp.Regexp, error) {
+	if re, ok := r.re[pattern]; ok {
+		return re, nil
+	}
+
+	re, err := regexp.Compile(pattern)
+	if err != nil {
+		return nil, err
+	}
+	r.re[pattern] = re
+	return re, nil
+}
+
+// CPETranslations include special case translations for software that fail to match entries in the NVD CPE Dictionary
+// using the standard logic. This may be due to unexpected vendor or product names.
+//
+// Example:
+//
+//     [
+//       {
+//         "match": {
+//           "bundle_identifier": ["com.1password.1password"]
+//         },
+//         "translation": {
+//           "product": ["1password"],
+//           "vendor": ["agilebits"]
+//         }
+//       }
+//     ]
+type CPETranslations []CPETranslationEntry
+
+func (c CPETranslations) Translate(reCache *regexpCache, s *fleet.Software) (CPETranslation, bool, error) {
+	for _, entry := range c {
+		match, err := entry.Match.Matches(reCache, s)
+		if err != nil {
+			return CPETranslation{}, false, err
+		}
+		if match {
+			return entry.Translation, true, nil
+		}
+	}
+
+	return CPETranslation{}, false, nil
+}
+
+type CPETranslationEntry struct {
+	Match       CPETranslationMatch `json:"match"`
+	Translation CPETranslation      `json:"translation"`
+}
+
+// CPETranslationMatch represents match criteria for cpe translations.
+type CPETranslationMatch struct {
+	Name             []string `json:"name"`
+	BundleIdentifier []string `json:"bundle_identifier"`
+	Source           []string `json:"source"`
+}
+
+// Matches returns true if the software satifies all the match criteria.
+func (c CPETranslationMatch) Matches(reCache *regexpCache, s *fleet.Software) (bool, error) {
+	match := func(a, b string) (bool, error) {
+		// check if its a regular expression enclosed in '/'
+		if len(a) > 2 && a[0] == '/' && a[len(a)-1] == '/' {
+			pattern := a[1 : len(a)-2]
```

Comment:
```go
pattern := a[1 : len(a)-1]
```

Reply:
Let's add some unit tests to `CPETranslationMatch.Matches`.

Reply:
nice catch

Reply:
There was a test `TestCPETranslations/match_name_regexp` but it was not a good enough test.